rs-matter/src/transport/core.rs

```diff
@@ -15,79 +15,198 @@
-    sx: [SxBuf; MAX_EXCHANGES + 1],
+const ACCEPT_TIMEOUT_MS: u64 = 1000;
+
+#[cfg(all(feature = "large-buffers", feature = "alloc"))]
+pub(crate) const MAX_RX_BUF_SIZE: usize = network::MAX_RX_LARGE_PACKET_SIZE;
+#[cfg(all(feature = "large-buffers", feature = "alloc"))]
+pub(crate) const MAX_TX_BUF_SIZE: usize = network::MAX_TX_LARGE_PACKET_SIZE;
+
+#[cfg(not(all(feature = "large-buffers", feature = "alloc")))]
+pub(crate) const MAX_RX_BUF_SIZE: usize = network::MAX_RX_PACKET_SIZE;
+#[cfg(not(all(feature = "large-buffers", feature = "alloc")))]
+pub(crate) const MAX_TX_BUF_SIZE: usize = network::MAX_TX_PACKET_SIZE;
+
+/// Represents the transport layer of a `Matter` instance.
+/// Each `Matter` instance has exactly one `TransportMgr` instance.
+///
+/// To the outside world, the transport layer is only visible and usable via the notion of `Exchange`.
+pub struct TransportMgr<'m> {
+    pub(crate) rx: IfMutex<NoopRawMutex, Packet<MAX_RX_BUF_SIZE>>,
+    pub(crate) tx: IfMutex<NoopRawMutex, Packet<MAX_TX_BUF_SIZE>>,
+    pub(crate) dropped: Notification<NoopRawMutex>,
+    pub session_mgr: RefCell<SessionMgr>, // For testing
+    pub(crate) mdns: MdnsImpl<'m>,
 }
 
-impl PacketBuffers {
-    const TX_ELEM: TxBuf = MaybeUninit::uninit();
-    const RX_ELEM: RxBuf = MaybeUninit::uninit();
-    const SX_ELEM: SxBuf = MaybeUninit::uninit();
-
-    const TX_INIT: [TxBuf; MAX_EXCHANGES] = [Self::TX_ELEM; MAX_EXCHANGES];
-    const RX_INIT: [RxBuf; MAX_EXCHANGES] = [Self::RX_ELEM; MAX_EXCHANGES];
-    const SX_INIT: [SxBuf; MAX_EXCHANGES + 1] = [Self::SX_ELEM; MAX_EXCHANGES + 1];
-
+impl<'m> TransportMgr<'m> {
     #[inline(always)]
-    pub const fn new() -> Self {
+    pub(crate) const fn new(mdns: MdnsImpl<'m>, epoch: Epoch, rand: Rand) -> Self {
         Self {
-            tx: Self::TX_INIT,
-            rx: Self::RX_INIT,
-            sx: Self::SX_INIT,
+            rx: IfMutex::new(Packet::new()),
+            tx: IfMutex::new(Packet::new()),
+            dropped: Notification::new(),
+            session_mgr: RefCell::new(SessionMgr::new(epoch, rand)),
+            mdns,
         }
     }
-}
 
-impl<'a> Matter<'a> {
+    #[cfg(all(feature = "large-buffers", feature = "alloc"))]
+    pub fn initialize_buffers(&self) -> Result<(), Error> {
+        let mut rx = self.rx.try_lock().map_err(|_| ErrorCode::InvalidState)?;
+        let mut tx = self.tx.try_lock().map_err(|_| ErrorCode::InvalidState)?;
+
+        if rx.buf.0.is_none() {
+            rx.buf.0 = Some(alloc::boxed::Box::new(heapless::Vec::new()));
+        }
+
+        if tx.buf.0.is_none() {
+            tx.buf.0 = Some(alloc::boxed::Box::new(heapless::Vec::new()));
+        }
+
+        Ok(())
+    }
+
+    #[cfg(not(all(feature = "large-buffers", feature = "alloc")))]
+    pub fn initialize_buffers(&self) -> Result<(), Error> {
+        // No-op, as buffers are allocated inline
+        Ok(())
+    }
+
+    pub fn reset(&self) {
+        self.session_mgr.borrow_mut().reset();
```

Comment:
This is probably a huge ask, however maybe we can plan on it:

- generally I would expect any unwrap() to have some comment explaining why it is safe and it will not panic
- RefCell borrow_mut and borrow can panic because they check rules at runtime.

Would it be appropriate to add some `safety` comment explaining why this kind of borrows would never panic? It may even need comments on the methods themselves.

Reply:
> This is probably a huge ask, however maybe we can plan on it:
> 
> * generally I would expect any unwrap() to have some comment explaining why it is safe and it will not panic

I'm with you here. Not commenting on `unwrap`s is probably my laziness and/or assuming things are obvious for everybody. Will (try to) comment the `unwrap`s introduced in the current PR, and maybe if you could open an issue for all unwraps? Or I can do this...

BTW: I was thinking we can use `expect` as you suggested elsewhere instead of `unwrap`, as `expect` leaves a trail at runtime too, but I have second thoughts on that - `expect` will blow up the `rodata` section isn't it? We don't want this for MCUs. `info!` `warn!` etc. are problematic enough already, but at least these are macros, so these can be compiled out.


> * RefCell borrow_mut and borrow can panic because they check rules at runtime.

Commenting every `RefCell` borrow would be like commenting every `Mutex` `lock` - because all `RefCell`s in this codebase are in fact either former `std::sync::Mutex` instances which I migrated to `RefCell`s when `no_std`-ifying the codebase, or newly introduced `Mutex`-es in disguise.

The reason why currently `RefCell` is used instead of a synchronous Mutex, is because there is no standard "Mutex" type in Rust _core_, and for single-threaded code a Mutex is essentially a `RefCell`.

In a single-threaded context, a `RefCell` will blow up at runtime exactly in the same conditions a mutex will - if you try to lock it twice within the same thread. So just like with mutexes, as long as the code borrowing the refcell is doing it for short spans and locally only, it would work.

So if we put a comment on `RefCell` borrows, it would be roughly the same everywhere, and would be something along the lines of "we only borrow the `RefCell` for a short code-span, and we have reviewed the code to not do double-borrow". Just like double-locking a mutex in the same thread (as mutexes are not reentrant).

I have plans for replacing all `RefCell`s with `embassy_sync::Mutex` instances, where `NoopMutex` or "a real raw mutex" is a decision at compile time, but I guess no rush, as the codebase would remain single-threaded for a long time to come.

> Would it be appropriate to add some `safety` comment explaining why this kind of borrows would never panic? It may even need comments on the methods themselves.

Maybe not exactly `safety` as `safety` concerns `unsafe` blocks in Rust and potential UB (RefCells and panics are not UB), but that's nitpicking and as per above I am with you on `unwrap`s.

For refcells, would be interested in your feedback!

P.S. forgot one dangerous use-case of RefCells (but exactly the same issue with _synchronous_ mutexes btw): you should **not** hold a borrow of a refcell across an await point. That's the one danger with refcells (and sync mutexes). Fortunately, Clippy has a warning for that.

Reply:
I changed my mind on `expect`. These blow up the `rodata` section. I have tried to comment all `unwrap`s in transport core now instead. If you find any other unwraps which are obvious, please let me know, we can comment these too.

Reply:
And as per above, I suggest we _don't_ comment all `RefCell` borrows and mutable borrows.